Implement Persistence Landscapes

catanzaromj opened the pull request:
This PR adds persistence landscapes with visualization and basic arithmetic methods. We implement two PL classes, one approximate for faster computations and one exact with no approximations. The approximate class is defined on a grid and we tried to use numpy vectorization when possible to make it fast, especially for things like linear combinations. There is a quicker 2D plotting method and a slower 3D variant as well. We also implemented L_p norms and transformer methods for the two classes.

sauln commented on persim/pers_landscape_visual.py:
```diff
@@ -0,0 +1,303 @@
+    ax.view_init(10, 90)
+    plt.show()
+
+def plot_landscape_exact_simple(landscape: PersLandscapeExact,
```
Could you run `black` over all the files?  This one looks a bit off.

catanzaromj replied:
Sure.

sauln commented on persim/pers_landscape_visual.py:
```diff
@@ -0,0 +1,303 @@
+    title=None,
+):
+    """
+    A plot of the exact persistence landscape.
```
Have you tested how the documentation site looks with these docstrings?  They look good from here, but sometimes the docs render a bit wonky.

catanzaromj replied:
Yes, but we can double check.

sauln commented on persim/__init__.py:
```diff
@@ -5,5 +5,10 @@
 from .visuals import *
+from .pers_landscape_exact import *
+from .pers_landscape_approx import *
+from .pers_landscape_transf import *
```
exact and approx make sense, but id k what `transf` and `aux` are for. Could you spell the whole words out here and in the filenames?

catanzaromj replied:
`transf` for transformer, where we implement scikit-learn style transformers, and `aux` for auxiliary. Should the transformers stay in their own file or should they move into their class files?

sauln commented on persim/pers_landscape_transf.py:
```diff
@@ -0,0 +1,69 @@
+__all__ = ["PLE", "PLA"]
+
+
+class PLE(BaseEstimator, TransformerMixin):
```
What does PLE stand for?

catanzaromj replied:
`PLE` is the PersLandscapeExact transformer, similarly for `PLA`. We're open to suggestions on re-naming.

sauln commented on persim/pers_landscape.py:
```diff
@@ -0,0 +1,72 @@
+import numpy as np
+
+
+class PersLandscape(ABC):
```
How is having the `PersLandscape` base class beneficial? It looks like in all the implementations, you implement `__init__` anyways.

catanzaromj replied:
It currently houses repeated code/checks from the two classes (the `__init__` calls `super()` in addition to other stuff), but it may have made more sense to have in an older version of our project. We can remove and refactor if you'd like.

sauln commented on persim/pers_landscape_approx.py:
```diff
@@ -0,0 +1,380 @@
+
+    Examples
+    --------
+
```
I would love to see some examples and notebooks showing these new implementations off!

catanzaromj replied:
For sure! We should be adding lots of examples in the next couple days.

sauln commented on persim/pers_landscape_approx.py:
```diff
@@ -0,0 +1,380 @@
+
+    def compute_landscape(self, verbose: bool = False) -> list:
+
+        verboseprint = print if verbose else lambda *a, **k: None
```
I love this :D

catanzaromj replied:
Awesome!

sauln commented on persim/pers_landscape_approx.py:
```diff
@@ -0,0 +1,380 @@
+
+
+############################################
+# End PersLandscapeApprox class definition #
```
Rather than having this delimiter, could the stuff below be moved into its own file?

catanzaromj replied:
Yes, or we could move it into the auxiliary file `pers_landscape_aux`.

sauln commented on persim/pers_landscape_approx.py:
```diff
@@ -0,0 +1,380 @@
+
+
+def snap_PL(
+    pls: list, start: float = None, stop: float = None, num_steps: int = None
```
I _love_ the typing by the way :D 

With this typing, can we do some static analysis of the code as well?  I'd love to incorporate that into our testing workflow if so.

catanzaromj replied:
Thanks. I'm not an expert on this, but I think some tools (like [mypy](https://mypy.readthedocs.io/en/stable/introduction.html)) allow you to do just that.

sauln commented on persim/pers_landscape_approx.py:
```diff
@@ -0,0 +1,380 @@
+"""
+    Persistence Landscape Approximate class
+
+    authors: Gabrielle Angeloro, Michael Catanzaro
```
How would you feel about adding your names to the `setup.py` authors list once rather than at the top of each of these files?

catanzaromj replied:
Sounds good.

sauln commented on persim/pers_landscape_approx.py:
```diff
@@ -0,0 +1,380 @@
+
+        # check if L is empty
+        if not L.size:
+            L = np.array(["empty"])
```
Is the string "empty" necessary, or can you just do `np.array([])`?

gabbyangeloro replied:
This is the case where the grid values entered miss all the birth death pairs, so the persistence landscapes are all zero. However, our algorithm will put the landscape as an empty list. 

We don't want the landscape (values) to be np.array([]) because then the algorithm won't know we already computed landscapes and will just keep computing them every time compute_landscape is called. 

I have pushed a change that puts L as a zeros array instead of np.array(["empty"]).

sauln commented on setup.py:
```diff
@@ -21,8 +21,11 @@
-      author_email='nat@riverasaul.com, chris.tralie@gmail.com, francis.c.motta@gmail.com',
+      author='Nathaniel Saul, Chris Tralie, Francis Motta, Michael Catanzaro,
+      Gabrielle Angeloro, Calder Sheagren',
+      author_email='nat@riverasaul.com, chris.tralie@gmail.com,
```
I looks like there's newlines between names and emails. I dont think that will work correctly with python strings.  What happens when you reinstall it with `pip install -e .`?

catanzaromj replied:
Yes, I'll fix this soon. Thanks for catching it.

sauln commented:
I've sent a PR to your fork (https://github.com/catanzaromj/persim/pull/3) with some minor structural changes and a bunch of docs updates. Let me know what you think.  

If you approve and merge that PR, I think the landscapes stuff is ready to be released!   Though it would be great to see some more tests for all the different utility functions and such.

Thanks for all your work implementing this, I'm very excited to get it incorporated :D

catanzaromj commented:
I think it looks really good. Thanks for your help with this last bit--the semester starting up has really spread my time thin. 

The only thing I noticed is the Persistence Landscapes and Machine Learning notebook isn't quite finished. @calderds do you have any final edits to make? I can also try to finish up the section 3 today if you're done.

calderds commented:
Mike, there's a version in the `calder` branch of your fork implementing the stuff we talked about, but with no increased performance. I can make some comments and put in a PR by this afternoon

catanzaromj commented:
Ok perfect. Ill have time to look at it and merge, and then I think we'll be all set for the final merge.

calderds commented:
I put in a PR to @catanzaromj 's fork with the last bit of documentation. Should be good to go on my end!

sauln commented:
Sweet!  I’ll get ready to merge and deploy everything this afternoon. Anything in particular you’d like me to call out when I make the Twitter announcement?

catanzaromj commented:
I think we're all set. I will add tests and work on tweaking the results of the ML notebook within the next month or so with a new PR. And I can't think of anything in particular for the announcement. Thanks for all your help with this!

sauln commented:
ahh actually I would prefer if we went ahead and added the tests before we release the code.  Ive found testing often uncovers bugs in the implementation, and once it's out, it's easy to never go back and add the tests. 

I'm okay letting this sit here a while till the tests are in place,  Thank you!

catanzaromj commented:
I totally get that. I will try to move it up on my list of things to do, but it will still probably be a couple weeks before everything is done.

catanzaromj commented:
@sauln I'm getting back to this, but I can't import the package.

```python
>>> import persim                                                                                                      
 Traceback (most recent call last):     
File "<stdin>", line 1, in <module>                                                                                     
File "E:\Miniconda\envs\baseenv\lib\site-packages\persim\__init__.py", line 8, in <module>                                    
from .landscapes import *                                                                                           
ModuleNotFoundError: No module named 'persim.landscapes'                                                                
```
I hope to have the additional tests added soon and I was hoping you could take care of this in the meantime.

calderds commented:
@catanzaromj  I got this error too and managed to fix it. I can push my fixes to the calder branch of your fork.

sauln commented:
Shoot, sorry about this! Thanks for taking care of it @calderds

catanzaromj commented:
I finished adding a bunch of tests and found a bug in the `p_norm` method, so it was definitely worth it. I added tests for pretty much everything except the visualization methods, since I don't really know how to test those. Feel free to add those if you'd like. I also did some minor renaming and refactoring, and updated the import statements on the examples given the new directory structure.

catanzaromj commented:
@sauln I think we're ready to go. Anything else that needs to be taken care of?

sauln commented on MANIFEST.in:
```diff
@@ -1,2 +1,3 @@
 include *.txt
 include *.md
+include persim/landscapes/*
```
This is surprising. Is this necessary? I thought setuptools will automatically include any `*.py` files.

sauln reviewed (approved):
Looks great! Thank you for putting all this work in 🙇   

One last thing before we merge:  Could you increment the version to 0.3.0 and add some notes to the `RELEASE.txt`?

catanzaromj commented:
@sauln Done. Thanks for all your help with this.

sauln commented:
Rad, let's ship it!

sauln commented:
Thank you @catanzaromj , @gabbyangeloro , and  @calderds !  I'm very glad to have an implementation of landscapes in the library finally.
